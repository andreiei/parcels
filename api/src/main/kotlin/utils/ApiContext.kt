package utils

import services.ParcelRepository
import services.RoutesService

data class ApiContext(
    val parcelRepository: ParcelRepository = ParcelRepository(),
    val routesService: RoutesService = RoutesService(),
)
