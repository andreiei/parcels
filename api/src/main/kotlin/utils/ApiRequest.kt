package utils

import com.amazonaws.services.lambda.runtime.events.APIGatewayProxyRequestEvent
import decodeFromString
import failure.ApiException
import kotlinx.serialization.Serializable

@Serializable
data class ApiRequest(
    val body: String?,
    val path: String,
    val httpMethod: String,
) {
    inline fun <reified T> getBody(): T {
        val rawBody: String = body ?: throw ApiException.BadRequestException("Request body is missing")
        return runCatching {
            decodeFromString<T>(rawBody)
        }
            .getOrElse {
                throw ApiException.BadRequestException("Failed to parse request body into ${T::class.simpleName}")
            }
    }

    fun getCustomPath(): String = "$httpMethod $path"

    companion object {
        fun APIGatewayProxyRequestEvent.toApiRequest(): ApiRequest = ApiRequest(
            body = body,
            path = path,
            httpMethod = httpMethod,
        )
    }
}
