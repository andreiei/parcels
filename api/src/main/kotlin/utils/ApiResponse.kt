package utils

import encodeToString
import kotlinx.serialization.Serializable

@Serializable
data class ApiResponse(
    val body: String?,
    val code: Int,
) {
    companion object {
        inline fun <reified T> withContent(body: T, code: Int = 200): ApiResponse =
            ApiResponse(body = encodeToString(body), code = code)

        fun noContent(): ApiResponse = ApiResponse(body = null, code = 204)
    }
}
