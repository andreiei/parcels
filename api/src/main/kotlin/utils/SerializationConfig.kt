import kotlinx.serialization.json.Json
import kotlinx.serialization.serializer

inline fun <reified T> encodeToString(data: T): String = SerializerConfig.json.encodeToString(serializer(), data)
inline fun <reified T> decodeFromString(data: String): T = SerializerConfig.json.decodeFromString(serializer(), data)

open class SerializerConfig {
    companion object {
        val json: Json =
            Json {
                isLenient = true
                prettyPrint = true
                encodeDefaults = true
                ignoreUnknownKeys = true
            }
    }
}
