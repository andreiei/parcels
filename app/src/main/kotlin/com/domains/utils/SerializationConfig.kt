package com.domains.utils

import kotlinx.serialization.json.Json
import kotlinx.serialization.serializer

inline fun <reified T> encodeToString(data: T): String = SerializerConfig.json.encodeToString(serializer(), data)

open class SerializerConfig {
    companion object {
        val json: Json =
            Json {
                isLenient = true
                encodeDefaults = true
                ignoreUnknownKeys = true
                prettyPrint = true
            }
    }
}
