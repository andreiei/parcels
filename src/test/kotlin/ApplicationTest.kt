import io.kotest.core.spec.style.ShouldSpec
import utils.shouldMatchSnapshot
import java.io.BufferedReader
import java.io.ByteArrayInputStream
import java.io.ByteArrayOutputStream
import java.io.InputStreamReader
import java.io.PrintStream

class ApplicationTest : ShouldSpec({
    should("handle empty input") {
        val inputString = "exit"

        val inputStream = ByteArrayInputStream(inputString.toByteArray())
        val reader = BufferedReader(InputStreamReader(inputStream))

        val outContent = ByteArrayOutputStream()
        val errContent = ByteArrayOutputStream()

        Application(
            input = reader,
            output = PrintStream(outContent),
            error = PrintStream(errContent),
        )
            .invoke()

        mapOf(
            "output" to outContent.toString().trim(),
            "error" to errContent.toString().trim(),
        )
            .shouldMatchSnapshot(this)
    }

    should("handle valid input") {
        val inputString = """
            fredrik@hotmail.com
            fredrik@hotmail.com ignorespace@hotmail.com
            malin@gmail.com

            test@domain..com
            exit
        """.trimIndent()

        val inputStream = ByteArrayInputStream(inputString.toByteArray())
        val reader = BufferedReader(InputStreamReader(inputStream))

        val outContent = ByteArrayOutputStream()
        val errContent = ByteArrayOutputStream()

        Application(
            input = reader,
            output = PrintStream(outContent),
            error = PrintStream(errContent),
        )
            .invoke()

        mapOf(
            "output" to outContent.toString().trim(),
            "error" to errContent.toString().trim(),
        )
            .shouldMatchSnapshot(this)
    }
})
