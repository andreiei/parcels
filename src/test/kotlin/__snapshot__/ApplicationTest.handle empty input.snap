{
    "output": "Enter one email address per line (type 'exit' to finish)\nNo email domains found.",
    "error": ""
}